### PR TITLE
Minor header fixes to reduce parsing issues with Clang

### PR DESCRIPTION
**Pull request by leoetlino**
A small set of fixes to libstratosphere headers to make stratosphere.hpp more parsable by Clang.

At the time of writing Clang has incomplete C++20 support so two of the commits add small, non-intrusive workarounds to allow Clang-based tools to still parse most of stratosphere.hpp and dump type definitions.

Unfortunately getting to 0 compilation errors is not possible at this time because Clang appears to have bugs in its constant evaluation engine and its definition of literal types does not conform to the C++17 standard. (That is the reason why the binary-search based offsetof implementation is broken in some cases on Clang.)

This PR is not just Clang workarounds though, it also fixes some non-standards-compliant code :)

### CODE REVIEW

File: libraries/libvapours/include/vapours/results/results_common.hpp [hunk 1]
```diff
@@ -177,7 +177,10 @@ namespace ams {
                 static_assert(Value != Base::SuccessValue, "Value != Base::SuccessValue");
             public:
                 constexpr ALWAYS_INLINE operator Result() const { return MakeResult(Value); }
-                constexpr ALWAYS_INLINE operator ResultSuccess() const { OnResultAbort(Value); }
+                constexpr ALWAYS_INLINE operator ResultSuccess() const {
+                    OnResultAbort(Value);
+                    return ResultSuccess();
```
Comment (SciresM) [inline libraries/libvapours/include/vapours/results/results_common.hpp:182]: I am very uncomfortable with this. OnResultAbort(Value) is NORETURN... Can __builtin_unreachable() or AMS_ASSUME(false) be used instead?

File: libraries/libstratosphere/include/stratosphere/sm/sm_types.hpp [hunk 2]
```diff
@@ -40,8 +40,8 @@ namespace ams::sm {
             return out;
         }
 
-        static constexpr ServiceName Encode(const char *name) {
-            return Encode(name, std::strlen(name));
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/sm/sm_types.hpp:44]: Wonder if we should just add e.g. util::Strlen. We have (as Nintendo does) Strnlen, Strlcpy...ah well, this is probably fine.

Comment (SciresM) [reply libraries/libstratosphere/include/stratosphere/sm/sm_types.hpp:44]: (no need to make changes here).

**Review by SciresM — CHANGES_REQUESTED** [review]
Pretty much just one change needed.

There's some minor style stuff too but I think it's all fine, nothing we don't already have in other places in the codebase.

File: libraries/libvapours/include/vapours/results/results_common.hpp (hunk 1 above)
Comment (leoetlino) [reply libraries/libvapours/include/vapours/results/results_common.hpp:183]: Added __builtin_unreachable() before the return. Unfortunately Clang requires a return statement to be present here even if the end of the function can never be reached...

**Pull request merged**